nyaa/api_torznab.py

```diff
@@ -101,71 +112,113 @@ def torznab_caps():
+
+            num = int(item)
+            if num >= 100000:
+                categories.append(num - 100000)
+            elif num >= 1000:
+                categories.extend([x - 100000 for x in TORZNAB_CATEGORY_MAP.get(num, [])])
+            elif num > 0:
+                categories.append(num)
+
+        # Check if any of the specified categories 'stuck'.
+        if not categories:
+            raise EmptySearchError('Categories {} guaranteed to produce empty result'.format(cat))
+    return categories
+
+def get_tag_filter(tag):
+    include_tags = []
+    exclude_tags = []
+    if tag:
+        for item in tag.split(','):
+            if item[0] == '!':
+                exclude_tags.append(item[1:].lower())
+            else:
+                include_tags.append(item.lower())
+    for incl in include_tags:
+        if incl in exclude_tags:
+            raise EmptySearchError('Tag {} both required and filtered')
+    return (include_tags, exclude_tags)
+
+def torznab_search(api_request):
+    page_size = TORZNAB_RESULTS_PER_PAGE
 
     # Default newznab query params, return empty result if unsupported to prevent clients from accessing the api as a traditional newznab indexer.
-    q = flask.request.args.get('q')
-    season = flask.request.args.get('season')
-    ep = flask.request.args.get('ep')
-    rid = flask.request.args.get('rid')
+    query_term = api_request.args.get('q')
+    query_season = api_request.args.get('season')
+    query_episode = api_request.args.get('ep')
+    query_rageid = api_request.args.get('rid')
 
-    cat = flask.request.args.get('cat', default='100000')
+    if query_season is not None or query_episode is not None or query_rageid is not None:
+        return render_torznab_error(201)
 
-    # Quality filter (remake/trusted) has no predefined query param yet. Propose 'tag' as generic cross newznab/torznab field to remain site agnostic.
-    #quality_filter = flask.request.args.get('f')
-    extended = flask.request.args.get('extended')
-    offset = flask.request.args.get('offset', default=0, type=int)
-    limit = flask.request.args.get('limit', default=page_size, type=int)
+    cat = api_request.args.get('cat', default='')
+    # Quality filter (remake/trusted) has no predefined query param yet.
+    # Propose 'tag' as generic cross newznab/torznab field to remain site agnostic.
+    tag = api_request.args.get('tag', default='')
+    extended = api_request.args.get('extended', False, type=bool)
+    offset = api_request.args.get('offset', default=0, type=int)
+    limit = api_request.args.get('limit', default=page_size, type=int)
 
-    if season or ep or rid:
-        return render_torznab_error(201)
+    try:
+        limit = min(limit, page_size)
+        if limit == 0:
+            raise EmptySearchError('Limit of 0 guaranteed guarantees empty result')
+        if offset % limit:
+            app.logger.warn('offset {} isn\'t a multiple of the page size {}'.format(offset, limit))
```

Comment:
Yes it throws when the offset isn't a multiple of the page size, that's what the warning literally says. Torznab allows you to set it mid-page, but it's rarely used.
Initially I had refactored search_elastic and search_db to use offset & limit instead of page & page_size. But that meant I couldn't use Paginator as was expected by parts of the UI code. And was getting increasingly complicated, I could've refactored that or have the legacy `search` create the paginator, but I felt that was more invasive and less likely to be accepted in the PR.
So instead I limited the torznab api to pages as 'acceptable' compromise.